asllib/doc/ASLTypingReference.tex

```diff
@@ -8694,2385 +8629,4792 @@ \subsection{Comments}
-      \item $\newtenv$ is $\tenvtwo$.
-    \end{itemize}
+    \item $\vx$ is not a member of $\vused$;
+    \item $\newtenv$ is $\tenvtwop$;
+    \item $\accp$ is $\acc$.
+  \end{itemize}
 
-    \item All of the following apply (\textsc{non\_contant}):
-    \begin{itemize}
-      \item $\keyword$ is not $\GDKConstant$;
-      \item $\newtenv$ is $\tenvone$.
-    \end{itemize}
+  \item All of the following apply (\textsc{used}):
+  \begin{itemize}
+    \item $\vx$ is a member of $\vused$;
+    \item checking that $\vx$ is not declared in $\tenvtwop$ yields $\True$\ProseOrTypeError;
+    \item annotating $\tty$ with identifier $\vx$ as a potential parameter type in $\tenvtwo$,
+          which is an environment to which all explicit parameters have been added but implicit parameters were not added to,
+          via $\annotateparamtype$ yields $\vt$\ProseOrTypeError;
+    \item checking whether $\vt$ is a constrained integer type in $\tenvtwo$ yields $\True$\ProseOrTypeError;
+    \item adding $\vx$ as a local storage element to $\tenvtwop$ with type $\vt$ and local declaration keyword $\LDKLet$ yields
+          $\newtenv$;
+    \item $\accp$ is $\acc$ updated by binding $\vx$ to $\vt$.
+  \end{itemize}
+\end{itemize}
+
+\begin{emptyformal}
+\subsection{Formally}
+\begin{mathpar}
+\inferrule[not\_used]{
+  \vx \not\in \vused
+}{
+  \argasparam(\tenvtwo, \vused, (\vx, \tty), (\tenvtwop, \acc))
+  \typearrow (\overname{\tenvtwop}{\newtenv}, \overname{\acc}{\accp})
+}
+\and
+\inferrule[used]{
+  \vx \in \vused\\
+  \checkvarnotinenv{\tenvtwop, \vx} \typearrow \True \OrTypeError\\\\
+  \annotateparamtype(\tenvtwo, \tty, \vx) \typearrow \vt \OrTypeError\\\\
+  \checkconstrainedinteger(\tenvtwo, \vt) \typearrow \True \OrTypeError\\\\
+  \addlocal(\tenvtwop, \vx, \vt, \LDKLet) \typearrow \newtenv\\
+  \accp \eqdef \acc[\vx\mapsto\vt]
+}{
+  \argasparam(\tenvtwo, \vused, (\vx, \tty), (\tenvtwop, \acc))
+  \typearrow (\newtenv, \accp)
+}
+\end{mathpar}
+\end{emptyformal}
+
+\section{TypingRule.AnnotateParamType \label{sec:TypingRule.AnnotateParamType}}
+\hypertarget{def-annotateparamtype}{}
+The function
+\[
+  \annotateparamtype(\overname{\staticenvs}{\tenv} \aslsep \overname{\ty}{\tty} \aslsep \overname{\identifier}{\vx})
+  \typearrow \overname{\ty}{\newty}
+\]
+annotates the type $\tty$ in $\tenv$, considering it as a subprogram parameter with identifier $\vx$,
+yielding the type $\newtenv$.
+It is assumed that $\tenv$ is an environment to which all explicitly defined parameters of the subprogram
+in context were added to, but the implicitly defined parameters were not added to.
+The result is a type error, if one is detected.
+
+\subsection{Prose}
+One of the following applies:
+\begin{itemize}
+  \item All of the following apply (\textsc{tint\_unconstrained}):
+  \begin{itemize}
+    \item $\tty$ is the unconstrained integer type;
+    \item $\newty$ is the underconstrained integer type for the identifier $\vx$.
```

Comment:
That's the term we currently use, until we change it to something like a parameterized integer type.
In this context, it may change to something "the integer type parameterzed by $\vx$".